wdl-cli/src/analysis/source.rs

```diff
@@ -0,0 +1,146 @@
+//! Sources for a WDL documents used in analysis.
+
+use std::path::PathBuf;
+
+use anyhow::Context as _;
+use anyhow::Result;
+use anyhow::anyhow;
+use anyhow::bail;
+use path_clean::clean;
+use url::Url;
+use wdl_analysis::Analyzer;
+use wdl_engine::path::parse_url;
+
+/// A source for an analysis.
+#[derive(Clone, Debug)]
+pub enum Source {
+    /// A remote URL.
+    Remote(Url),
+
+    /// A local file.
+    File(Url),
+
+    /// A local directory.
+    Directory(PathBuf),
+}
+
+impl Source {
+    /// Attempts to reference the source as a URL.
+    pub fn as_url(&self) -> Option<&Url> {
+        match self {
+            Source::Remote(url) | Source::File(url) => Some(url),
+            Source::Directory(_) => None,
+        }
+    }
+
+    /// Registers the source within an [`Analyzer`].
+    pub async fn register<T: Send + Clone + 'static>(
+        self,
+        analyzer: &mut Analyzer<T>,
+    ) -> Result<()> {
+        match self {
+            Source::Remote(url) | Source::File(url) => analyzer.add_document(url).await,
+            Source::Directory(path) => analyzer.add_directory(path).await,
+        }
+    }
+}
+
+impl std::fmt::Display for Source {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Source::Remote(url) => write!(f, "{url}"),
+            Source::File(url) => write!(f, "{url}"),
+            Source::Directory(path) => write!(f, "{path}", path = path.display()),
+        }
+    }
+}
+
+impl std::str::FromStr for Source {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some(url) = parse_url(s) {
```

Comment:
This is deserving of being called out. @peterhuene, since there is no great way to tell if a string is a remote URL or a local file/directory, I've relied on your function in `wdl-engine` to check that. That's the best solution I can think of without purely duplicating the function, and at least it means that `wdl-engine` and `wdl-cli` will be in sync. That being said, it does require the `path` module of `wdl-engine` to be made public.